### PR TITLE
fix: Composition with sawtooth

### PR DESCRIPTION
**Pull request by ryanbtp**
Chronicle base upgrade and significant improvements to build times

* 0.6 required to run devmode without using compose and a pg image
* Fix e2e with validator
* Ensure correct version of tp is used
* Drop chronicle to non root user in docker - required for embedded PG
* Add ca-certs for pg-embed download
* Introduce buildx for incremental compilation, for faster iteration on domain development
* Cache layer dependent on FEATURES, Release / Debug and builder version
* /app/target cached for dependencies
* Domain build forces a rebuild from domain.yaml by touching it
* make clean now removes ctx, if non incremental compilation is required
* Clean up 'stl' -> sawtooth as it's confusing terminology
* build target now compiles debug sdl for all domains - suitable for CI

References: BCP-21

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -21,8 +21,8 @@ a good starting point is [Introduction to GraphQL](https://graphql.org/learn/).
 git clone https://github.com/btpworks/chronicle-examples.git
 ```
 
-This contains several example domain yaml files, and Docker uses
-`blockchaintp/chronicle-builder:BTP2.1.0` as the builder image by default.
+This contains several example domain yaml files and docker and uses
```
Comment (mtbc) [inline README.md:24]: It contains docker?

File: docker/chronicle-test/chronicle-test [hunk 2]
```diff
@@ -0,0 +1,34 @@
+#!/usr/bin/env bash
+
+echo "Wait for stack"
```
Comment (mtbc) [inline docker/chronicle-test/chronicle-test:3]: Could `set -e` be a safe/good option?

File: README.md (hunk 1 above)
Comment (ryanbtp) [reply README.md:24]: Oops

**Comment by mtbc** [comment]
This seems to run fine on x86_64 GNU/Linux.

File: docker/chronicle.dockerfile [hunk 3]
```diff
@@ -25,16 +51,25 @@ RUN if [ "${RELEASE}" = "yes" ]; then \
   fi;
 
 WORKDIR /
-FROM ubuntu:focal AS domain
+FROM ubuntu:focal AS domain-base
```
Comment (suchapalaver) [inline docker/chronicle.dockerfile:54]:
```suggestion
FROM ubuntu:jammy AS domain-base
```
That change got things working for me, working off this branch.

File: Makefile [hunk 4]
```diff
@@ -58,89 +67,135 @@ else
...
+.PHONY: build-end-to-end-test
+
+.PHONY: build-end-to-end-test
+build-end-to-end-test:
+	docker build -t chronicle-test:$(ISOLATION_ID) -f docker/chronicle-test/chronicle-test.dockerfile .
+
+.PHONY: test-e2e
+test-e2e: build-end-to-end-test
+	COMPOSE_PROFILES=test CHRONICLE_IMAGE=chronicle-evidence-sawtooth CHRONICLE_VERSION=$(ISOLATION_ID) \
+	 $(COMPOSE) -f docker/chronicle.yaml up --exit-code-from chronicle-test
```
Comment (suchapalaver) [inline Makefile:177]:
```suggestion
	 $(COMPOSE) -f docker/docker-compose.yaml up --exit-code-from chronicle-test
```

**Comment by suchapalaver** [comment]
Running `make run-sawtooth-manufacturing`, I still get the following error:

```
export CHRONICLE_IMAGE=chronicle-manufacturing-sawtooth; export CHRONICLE_VERSION=local; docker-compose -f docker/chronicle.yaml up -d
ERROR: .FileNotFoundError: [Errno 2] No such file or directory: './docker/chronicle.yaml'
make: *** [Makefile:207: run-sawtooth-manufacturing] Error 1
```

File: Makefile [hunk 5]
```diff
@@ -20,35 +23,41 @@ clean: clean_containers
...
 
-build: all-domains
-
 clean_containers:
-	docker-compose -f docker/docker-compose.yaml rm -f || true
+	docker-compose -f docker/chronicle.yaml rm -f || true
 
 clean_docker:
-	docker-compose -f docker/docker-compose.yaml down -v --rmi all || true
+	docker-compose -f docker/cronicle.yaml down -v --rmi all || true
```
Comment (suchapalaver) [inline Makefile:30]:
```suggestion
	docker-compose -f docker/chronicle.yaml down -v --rmi all || true
```

**Comment by suchapalaver** [comment]
Trying this out, I've had things working successfully with the logs showing things working in the validator and the domain container, but I also keep running into the following error:

```
Pulling domain (chronicle-manufacturing-sawtooth:local)...
ERROR: The image for the service you're trying to recreate has been removed. If you continue, volume data could be lost. Consider backing up your data before continuing.

Continue with the new image? [yN]y
Pulling domain (chronicle-manufacturing-sawtooth:local)...
ERROR: pull access denied for chronicle-manufacturing-sawtooth, repository does not exist or may require 'docker login': denied: requested access to the resource is denied
make: *** [Makefile:211: run-sawtooth-manufacturing] Error 1
```
Building everything again from scratch the `make run-sawtooth-manufacturing` process stops at the above error. That image does not exist for me, only the following:

```
$ docker images
REPOSITORY                                  TAG              IMAGE ID       CREATED         SIZE
blockchaintp/chronicle-tp-amd64             BTP2.1.0-0.6.0   b9a14875b161   6 days ago      101MB
postgres                                    13.9-alpine      8e750948d39a   4 weeks ago     238MB
blockchaintp/sawtooth-validator             BTP2.1.0         42a6dfdbfc22   7 weeks ago     749MB
blockchaintp/sawtooth-settings-tp           BTP2.1.0         0202fb5b0aaf   7 weeks ago     223MB
blockchaintp/sawtooth-rest-api              BTP2.1.0         e0e708a4b477   7 weeks ago     272MB
blockchaintp/sawtooth-devmode-engine-rust   BTP2.1.0         dbd8b60f760a   13 months ago   300MB
```

File: Makefile (hunk 5 above)
Comment (mtbc) [reply Makefile:30]: Still seems to need fixed?

File: Makefile [hunk 6]
```diff
@@ -58,89 +67,137 @@ else
...
+.PHONY: build-end-to-end-test
+build-end-to-end-test:
+	docker build -t chronicle-test:$(ISOLATION_ID) -f docker/chronicle-test/chronicle-test.dockerfile .
+
+.PHONY: test-e2e
+test-e2e: build-end-to-end-test
+	COMPOSE_PROFILES=test CHRONICLE_IMAGE=chronicle-evidence-sawtooth CHRONICLE_VERSION=$(ISOLATION_ID) \
+	 $(COMPOSE) -f docker/chronicle.yaml up --exit-code-from chronicle-test
+
+.PHONY: package
```
Comment (scealiontach) [inline Makefile:181]: package analyze publish and test are all PHONYed and defined in standard_defs.mk

File: Makefile [hunk 7]
```diff
@@ -58,89 +67,137 @@ else
...
 		--build-arg CHRONICLE_BUILDER_IMAGE=$(CHRONICLE_BUILDER_IMAGE) \
 		--build-arg RELEASE=no \
 		--build-arg FEATURES="" \
-		--build-arg DOMAIN=$(1)
+		--build-arg DOMAIN=$(1) . \
+		--load
 	@$(DOCKER_TAG) chronicle-domain:$(ISOLATION_ID) \
-		chronicle-$(1)-stl:$(ISOLATION_ID)
-	@touch $(MARKERS)/$@
+		chronicle-$(1)-sawtooth:$(ISOLATION_ID)
```
[the remark this answers is not in the document]
Comment (ryanbtp) [reply Makefile:104]: The reason was me getting confused between stl and sdl. I'll change it back

File: docker/chronicle.dockerfile [hunk 8]
```diff
@@ -1,14 +1,40 @@
-ARG CHRONICLE_BUILDER_IMAGE=blockchaintp/chronicle-builder
-ARG CHRONICLE_VERSION=BTP2.1.0
-FROM ${CHRONICLE_BUILDER_IMAGE}:${CHRONICLE_VERSION} as builder
+# syntax=docker/dockerfile:1.4
+ARG CHRONICLE_BUILDER_IMAGE
+ARG CHRONICLE_VERSION
```
[the remark this answers is not in the document]
Comment (ryanbtp) [reply docker/chronicle.dockerfile:3]: Now has the architecture in the name of the image

**Comment by ryanbtp** [comment]
Rebuild after cleaning containers without cleaning the markers should now work, with a small cost to build times

**Comment by mtbc** [comment]
Out of curiosity I tried `make run-stl-artworld` and it started,
```
...
Makefile:183: warning: overriding recipe for target 'build-end-to-end-test'
Makefile:183: warning: ignoring old recipe for target 'build-end-to-end-test'
Makefile:183: warning: overriding recipe for target 'test-e2e'
Makefile:183: warning: ignoring old recipe for target 'test-e2e'
make: Circular artworld-stl <- artworld-stl dependency dropped.
...
```
(It looked fairly happy afterward.)

**Comment by mtbc** [comment]
Build runs into,
> error: couldn't read crates/chronicle-domain/src/main.rs: No such file or directory

**Comment by ryanbtp** [comment]
> Build runs into,
> 
> > error: couldn't read crates/chronicle-domain/src/main.rs: No such file or directory

Fixed, hopefully! CI test was using wrong entrypoint

File: docker/chronicle.dockerfile (hunk 8 above)
[the remark this answers is not in the document]
Comment (scealiontach) [reply docker/chronicle.dockerfile:3]: Should still have a default here.  Put the default back in with the amd64 image or if you like make it smarter with TARGETARCH

File: docker/chronicle.yaml [hunk 9]
```diff
@@ -19,29 +20,20 @@ services:
...
-            sawtooth.consensus.min_wait_time=0 \
-            sawtooth.consensus.algorithm.name=devmode \
-            sawtooth.consensus.algorithm.version=1.0 \
-            sawtooth.consensus.max_wait_time=0 -o consensus-settings.batch
-          sawadm genesis config-genesis.batch consensus-settings.batch
-        fi
-        sawtooth-validator -vvvv --scheduler parallel \
-          --endpoint tcp://validator:8800 --bind component:tcp://eth0:4004 \
-          --bind consensus:tcp://eth0:5050 --bind network:tcp://eth0:8800
+    entrypoint: >
```
Comment (scealiontach) [inline docker/chronicle.yaml:23]: Can we put the formatting back to the original here.  The `>` and `&&` version is both less readable and more error-prone, than the `-exc` and `|` version.

File: docker/chronicle.yaml [hunk 10]
```diff
@@ -79,23 +71,42 @@ services:
...
+    environment:
+      - RUST_BACKTRACE=full
+      - RUST_LOG=trace
+      - PGHOST=chronicle-pg
+      - PGPASSWORD=chronicle
+      - PGUSER=chronicle
+    image: &chronicleImage ${CHRONICLE_IMAGE}:${CHRONICLE_VERSION}
+    entrypoint: "/bin/bash"
+    command:
+      - -xc
```
Comment (scealiontach) [inline docker/chronicle.yaml:93]: if you make it  `-exc` you can lose the `&&`'s

**Review by scealiontach — CHANGES_REQUESTED** [review]

**Review by scealiontach — APPROVED** [review]

**Pull request merged**